source/dol/mLib/m_mtx.cpp

```diff
@@ -0,0 +1,177 @@
+    mData[2][0] = _20;
+    mData[2][1] = _21;
+    mData[2][2] = _22;
+    mData[2][3] = _23;
+}
+
+void mMtx_c::XrotS(mAng angle) {
+    float cos = nw4r::math::CosIdx((s16)angle.mAngle);
+    register float sin = nw4r::math::SinIdx((s16)angle.mAngle);
+
+    mData[0][0] = 1.0f;
+
+    // [The fneg instruction is generated at the wrong place no matter what.
+    // Since I couldn't find a way to overcome this issue, I just forced CW's hand].
+    #ifndef NONMATCHING
+
+    register float negSin;
+    asm volatile {
+        fneg negSin, sin;
+    }
+    #endif
+
+    float zero = 0.0f;
+    mData[0][1] = zero;
+    mData[0][2] = zero;
+    mData[0][3] = zero;
+    mData[1][0] = zero;
+    mData[1][1] = cos;
+
+    // [Follow-up from previous non-matching code].
+    #ifdef NONMATCHING
+    mData[1][2] = -sin;
+    #else
+    mData[1][2] = negSin;
+    #endif
+
+    mData[1][3] = zero;
+    mData[2][0] = zero;
+    mData[2][1] = sin;
+    mData[2][2] = cos;
+    mData[2][3] = zero;
+}
+
+void mMtx_c::XrotM(mAng angle) {
+    if (angle.mAngle != 0) {
+        mMtx_c rotatedMtx = mMtx_c();
+        rotatedMtx.XrotS(angle);
+        PSMTXConcat((const Mtx *)this, (const Mtx *)&rotatedMtx, (Mtx *)this);
+    }
+}
+
+void mMtx_c::YrotS(mAng angle) {
+    float cos = nw4r::math::CosIdx((short)angle.mAngle);
+    float sin = nw4r::math::SinIdx((short)angle.mAngle);
+    mData[0][0] = cos;
+    mData[0][1] = 0.0f;
+    mData[0][2] = sin;
+    mData[0][3] = 0.0f;
+    mData[1][0] = 0.0f;
+    mData[1][1] = 1.0f;
+    mData[1][2] = 0.0f;
+    mData[1][3] = 0.0f;
+    mData[2][0] = -sin;
+    mData[2][1] = 0.0f;
+    mData[2][2] = cos;
+    mData[2][3] = 0.0f;
+}
+
+void mMtx_c::YrotM(mAng angle) {
+    if (angle.mAngle != 0) {
+        mMtx_c rotatedMtx = mMtx_c();
+        rotatedMtx.YrotS(angle);
+        PSMTXConcat((const Mtx*)this, (const Mtx*)&rotatedMtx, (Mtx*)this);
+    }
+}
+
+void mMtx_c::ZrotS(mAng angle) {
+    float cos = nw4r::math::CosIdx((short)angle.mAngle);
+    float sin = nw4r::math::SinIdx((short)angle.mAngle);
+
+    float one = 1.0f;
+    float zero = 0.0f;
+    float negSin = -sin;
+
+    mData[0][0] = cos;
+    mData[0][2] = zero;
+    mData[0][1] = negSin;
+    mData[0][3] = zero;
+    mData[1][0] = sin;
+    mData[1][1] = cos;
+    mData[1][2] = zero;
+    mData[1][3] = zero;
+    mData[2][0] = zero;
+    mData[2][1] = zero;
+    mData[2][2] = one;
+    mData[2][3] = zero;
+}
+
+void mMtx_c::ZrotM(mAng angle) {
+    if (angle.mAngle != 0) {
+        mMtx_c rotatedMtx = mMtx_c();
+        rotatedMtx.ZrotS(angle);
+        PSMTXConcat((const Mtx*)this, (const Mtx*)&rotatedMtx, (Mtx*)this);
+    }
+}
+
+void mMtx_c::ZXYrotM(mAng xRot, mAng yRot, mAng zRot) {
+    YrotM(yRot);
+    XrotM(xRot);
+    ZrotM(zRot);
+}
+
+void mMtx_c::XYZrotM(mAng xRot, mAng yRot, mAng zRot) {
+    ZrotM(zRot);
+    YrotM(yRot);
+    XrotM(xRot);
+}
+
+inline float calcLength(float x, float y) {
+    x *= x;
+    y *= y;
+    x += y;
+    return x;
```

Comment:
I'm guessing this is the way it is because it wouldn't match otherwise? `return x * x + y * y;` is much clearer but I assume that won't match?

Reply:
Yep, won't match.